base/genericmemory.jl

```diff
@@ -0,0 +1,261 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+## genericmemory.jl: Managed Memory
+
+"""
+    GenericMemory{isatomic, T} <: AbstractVector{T}
+
+One-dimensional dense array with elements of type `T`.
+"""
+GenericMemory
+"""
+    Memory{T} == GenericMemory{:not_atomic, T}
```

Comment:
It would be nice to have a more formal representation of accessibility and mutability. Maybe a formal type that we could attach documentation to and dispatch on, if necessary. I understand that we don't want to overcommit to a hierarchy of types since the meaning could change based on context (annotating a function vs a stored object). But even something like this might be useful.

```julia
baremodule MemoryKinds
    struct Constant end
    baremodule Mutable
        struct NotAtomic end
        struct Unordered end
        struct Monotonic end
        struct Acquire end
        struct Release end
        struct AcquireRelease end
        struct SequentiallyConsistent end
    end
end
```

Reply:
so the currently expected options are `not_atomic` and `atomic`. For `atomic` memories, the kind of memory order is a property of the `setindex!`, not of the datatype.

Reply:
> For `atomic` memories, the kind of memory order is a property of the `setindex!`, not of the datatype.

I'm not currently seeing any code for the `atomic` variant of `setindex!`. Do you mean that the exact order in which memory is accessed is specified when calling `setindex!`, similar to how `unsafe_store!` works?

I thought there was talk of also having a constant variant for that parameter and figured this would provide explicit support in the type for what kind of access is supported. So the atomic variant would be `GenericMemory{Union{Unordered, Monotonic, Acquire, Release, AcquireRelease, SequentiallyConsistent}, T}`, the current implementation would be `GenericMemory{NotAtomic, T}`, and the constant variant would be `GenericMemory{Constant, T}`.

We could also just punt the full specification down the line a bit, but if we do then we should probably remove explicit mention of that field so it's not erroneously considered a stable interface that people should mess with.

Reply:
as mentioned in the design doc, most of the code for atomic `GenericMemory` is not written yet. We just need to know what the design will be so we don't discover in the future that we need an extra type parameter that would be breaking to add
> Do you mean that the exact order in which memory is accessed is specified when calling setindex!, similar to how unsafe_store! works?

yes.

Reply:
Apologies, I wasn't aware that the design doc was still undergoing revisions at this point. Would it be better to comment there since this detail may not make it into this PR?

Reply:
The primary argument against making the kind type parameter a proper type was that there hardly is any hierarchy between memory kinds. They are (mostly) mutually exclusive, and so using a symbol for them is usually fine. Not being able to add docstrings to those symbols is a bit of an argument, but something that could be solved by

```
const NotAtomic = :not_atomic
```

too, where the docstring would then be attached to `NotAtomic`. Personally, I dislike this kind of magic-constantness (Ref. https://github.com/JuliaLang/julia/issues/51357) because it hides the fact that this object has a special semantic meaning on the value/type level - you're not supposed to use that object outside of the narrow context of memory kinds in `GenericMemory` after all. 

Maybe this is on example where being able to specify an `isa` relationship in `where` clauses would be beneficial, in order to ensure that not just any random symbol is accepted as a memory kind.

Reply:
The main benefit really comes from being able to dispatch on kinds (as discussed in the [JuliaAtomics proposal](https://gist.github.com/vtjnash/11b0031f2e2a66c9c24d33e810b34ec0#alternatives-considered)) and test the validity of a kind of memory access.

For pointers and properties/fields, this is mostly a matter of convenience since we could probably do some compiler work to ensure symbols and modifying functions propagate appropriately. However, the array interface is a lot messier since `getindex(kind, a::AbstractArray, inds...)` may have to propagate `kind` through a number of wrapping structures before arriving at the innermost type and being tested for validity.

Reply:
The JuliaAtomics manifesto has so many TODOs/things that must have been decided by now due to being required for the implementation that I really don't think pointing to it as a source of truth is helpful (ref. https://github.com/JuliaLang/julia/issues/46739). I also don't see any reference to dispatching on kind types; those are always symbols there.